sn_networking/src/event.rs

```diff
@@ -652,25 +654,73 @@ impl SwarmDriver {
+                        info!("Query task {id:?} QuorumFailed record {pretty_key:?} among peers {peers:?} with quorum {quorum:?}, {stats:?} - {step:?}");
                     }
                     GetRecordError::Timeout { key } => {
-                        info!(
-                            "Query task {id:?} timed out when looking for record {:?}",
-                            PrettyPrintRecordKey::from(key.clone())
+                        let pretty_key = PrettyPrintRecordKey::from(key.clone());
+
+                        debug!(
+                            "Query task {id:?} timed out when looking for record {pretty_key:?}"
                         );
+
+                        let (sender, result_map, quorum, _expected_holders) =
+                            self.pending_get_record.remove(&id).ok_or_else(|| {
+                                trace!(
+                                    "Can't locate query task {id:?} for {pretty_key:?}, it has likely been completed already."
+                                );
+                                Error::ReceivedKademliaEventDropped( KademliaEvent::OutboundQueryProgressed {
+                                    id,
+                                    result: QueryResult::GetRecord(Err(err.clone())),
+                                    stats,
+                                    step,
+                                })
+                            })?;
+
+                        let required_response_count = match quorum {
+                            GetQuorum::Majority => close_group_majority(),
+                            GetQuorum::All => CLOSE_GROUP_SIZE,
+                            GetQuorum::One => 1,
+                        };
+
+                        // if we've a split over the result xorname, then we don't attempt to resolve this here.
+                        // Retry and resolve through normal flows without a timeout.
+                        if result_map.len() > 1 {
+                            warn!("Get record task {id:?} for {pretty_key:?} timed out with split result map");
+                            sender
+                                .send(Err(Error::QueryTimeout))
+                                .map_err(|_| Error::InternalMsgChannelDropped)?;
+                            return Ok(());
+                        }
+
+                        // if we have enough responses here, we can return the record
+                        if let Some((record, peers)) = result_map.values().next() {
+                            if peers.len() >= required_response_count {
+                                sender
+                                    .send(Ok(record.clone()))
+                                    .map_err(|_| Error::InternalMsgChannelDropped)?;
+                                return Ok(());
+                            }
+                            debug!("Get record task {id:?} for {pretty_key:?} returned insufficient responses")
```

Comment:
better log the non-responded `expected_holders` here as well ?

Reply:
the responded expected node will be removed from the `expected_holders` list.
hence here only need to add `{expected_holders:?}` into the `debug!()` log.